### PR TITLE
Prevent IndexShipper from leaking mutex accesses.

### PR DESCRIPTION
**Pull request by owen-d**
# What

Inverts control in the `index-shipper` so consumers don't manage accesses to mutexes. This prevents deadlocks and no longer requires consumers have internal knowledge of the `index-shipper` itself.

This PR reverts a few previous attempts. Look at the last commit to see the net-additions.

# Background:

TSDB previously didn't use mmap, which meant we could freely pass around TSDBs as `[]byte`. Since they were immutable, we didn't need to know when an underlying index was cleaned up/closed; this coordination was unnecessary and the garbage collector would clean up dangling memory for us. After introducing TSDB, we needed to control access to avoid use-after-free bugs. A few attempts at controlling this became increasingly complex & error prone. Instead of requiring _consumers_ to prevent deadlocks on index-shipper internals, it makes more sense that references to the index-shipper's resources should not leak outside the `ForEach` functions that give temporary access to them. This _inverts control_ from the consumer (code which queries the indices) to the producer (the index-shipper).

### CODE REVIEW

**Comment by periklis** [comment]
@owen-d Is the leackage serious enough to add the backport label for 2.7.0?

File: pkg/storage/stores/indexshipper/downloads/table.go [hunk 1]
```diff
@@ -145,24 +146,48 @@ func (t *table) Close() {
...
-	sets, err := t.getOrCreateIndexSets(ctx, ids, true)
-	if err != nil {
-		return err
+func (t *table) ForEachConcurrent(ctx context.Context, userID string, callback index.ForEachIndexCallback) error {
+	g, ctx := errgroup.WithContext(ctx)
+
+	// iterate through both user and common index
+	users := []string{userID, ""}
+
+	for i := range []string{userID, ""} {
```
Comment (DylanGuedes) [inline pkg/storage/stores/indexshipper/downloads/table.go:155]:
```suggestion
	for i := range users {
```

File: pkg/storage/stores/tsdb/index_shipper_querier.go [hunk 2]
```diff
@@ -29,36 +31,35 @@ func newIndexShipperQuerier(shipper indexShipperIterator, tableRanges config.Tab
...
+				if !ok {
+					return fmt.Errorf("unexpected shipper index type: %T", idx)
+				}
+				if multitenant {
+					impl = NewMultiTenantIndex(impl)
+				}
+
+				return f(ctx, impl)
+			}); err != nil {
+				return err
```
Comment (DylanGuedes) [inline pkg/storage/stores/tsdb/index_shipper_querier.go:56]: i think here you're allowing `itr` to be of type `err`. I'm not sure the NewMultiIndex at line 62 will be able of handling it though :thinking:

Comment (sandeepsukhani) [reply pkg/storage/stores/tsdb/index_shipper_querier.go:56]: The return here is part of the function definition cast to `indexIterFunc`. It looks fine to me. `itr` is holding a function definition and not capturing the output of a function call.

**Comment by sandeepsukhani** [comment]
> @owen-d Is the leackage serious enough to add the backport label for 2.7.0?

the code that we are fixing here never made it to any release so not required.

**Review by sandeepsukhani — APPROVED** [review]
Let us leave a note somewhere on lazy iteration possibly going over different index files within the same index query. For e.g, `Bounds` and `GetChunkRefs` might go through different index files if a sync happened between the calls.

The code otherwise LGTM.

File: pkg/storage/stores/tsdb/index_shipper_querier.go (hunk 2 above)
Comment (DylanGuedes) [reply pkg/storage/stores/tsdb/index_shipper_querier.go:56]: ohhh you right, thank you

**Pull request merged**